src/block.rs

```diff
@@ -1,60 +1,68 @@
+pub struct Blocks(pub Vec<Block>);
 
-    /// Determine an overlap with another block
-    ///
-    /// Convenience function to determine overlaps between sequence regions.
-    ///
-    /// Note: Blocks of the same type should usually not overlap in practise.
-    pub fn overlap(&self, other: &Block) -> Option<Range<usize>> {
-        let start = max(self.start, other.start);
-        let end = min(self.start + self.length, other.start + other.length);
-        if start < end {
-            Some(start as usize..end as usize)
-        } else {
-            None
-        }
+impl Deref for Blocks {
+    type Target = [Block];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl Blocks {
+    #[inline]
+    pub fn iter_overlaps(&self, range: impl RangeBounds<usize>) -> impl Iterator<Item = &Block> {
+        let clone_bound = |bound: Bound<&usize>| {
+            // stable in Rust 1.55 but not in 1.51
+            match bound {
+                Bound::Unbounded => Bound::Unbounded,
+                Bound::Included(x) => Bound::Included(*x),
+                Bound::Excluded(x) => Bound::Excluded(*x),
+            }
+        };
+        let start = clone_bound(range.start_bound());
+        let end = clone_bound(range.end_bound());
+        self.iter()
+            .skip_while(move |block| match start {
+                Bound::Included(v) => block.end <= v,
+                Bound::Excluded(v) => block.end <= v + 1,
+                Bound::Unbounded => false,
+            })
+            .take_while(move |block| match end {
+                Bound::Included(v) => block.start < v + 1,
+                Bound::Excluded(v) => block.start < v,
+                Bound::Unbounded => true,
+            })
+    }
 
-    #[test]
-    fn test_overlap() {
-    let a = Block::new(5, 10);
-    let b = Block::new(10, 10);
-    let c = Block::new(15, 10);
-    let d = Block::new(100, 25);
-    let e = Block::new(90, 40);
-    assert_eq!(a.overlap(&b), Some(10 .. 15));
-    assert_eq!(a.overlap(&c), None);
-    assert_eq!(b.overlap(&c), Some(15 .. 20));
-    assert_eq!(a.overlap(&a), Some(5 .. 15));
-    assert_eq!(a.overlap(&d), None);
-    assert_eq!(d.overlap(&a), None);
-    assert_eq!(d.overlap(&e), Some(100..125));
-    assert_eq!(e.overlap(&d), Some(100..125));
-    assert_eq!(Block::new(67920552, 300000).overlap(&Block::new(67859142, 61415 )), Some(67920552 .. (67859142 + 61415)));
+    #[must_use]
+    pub fn count(&self) -> usize {
+        self.iter().fold(0, |acc, block| acc + block.len())
+    }
+
+    pub fn apply_masks<const HARD: bool>(&self, seq: &mut [u8], range: Block) {
+        let (start, end) = (range.start, range.end);
+        for block in self.iter_overlaps(range) {
+            let seq_start = start.max(block.start) - start;
+            let seq_end = end.min(block.end) - start;
+            for i in seq_start..seq_end {
+                unsafe {
```

Comment:
There should be a comment above this line that says something like `// safe because for-loop ensures valid bounds `

Reply:
I'm actually doubt that this is really safe. The parameter `seq` does not get cross-checked by anything.

Reply:
I think the way to make it safe is to just pass `start` instead of `range`.

Because the implicit invariant here now is that `seq.len() == range.end - range.start`.